designsafe/apps/api/datafiles/views.py

```diff
@@ -82,6 +83,24 @@ def get(self, request, api, operation=None, scheme='private', system=None, path=
             return JsonResponse(response)
         except (BoxOAuthException, DropboxAuthError, GoogleAuthError):
             raise resource_expired_handler(api)
+        except (InternalServerError, UnauthorizedError) as e:
+            error_status = e.response.status_code
+            if error_status in [401, 500]:
+                logger.info(e)
+
+                system_needs_keys = test_system_needs_keys(client, request.user.username, system)
+                if system_needs_keys:
+                    logger.info(
+                        f"Keys for user {request.user.username} must be manually pushed to system: {system_needs_keys.id}"
+                    )
+                    return JsonResponse({'message': str(e)}, status=e.response.status_code)
+
+                # If the user has valid system credentials, retry the request
```

Comment:
Instead of duplicating the view code here, can we raise a new exception to be caught by the `@retry` decorator?